src/variables/utils/getWindowVars.ts

```diff
@@ -2,152 +2,58 @@
-
-  return [
-    {
-      type: 'VariableAssignment',
-      id: {
-        type: 'Identifier',
-        name: WINDOW_PERIOD,
-      },
-      init: {
-        type: 'DurationLiteral',
-        values: [{magnitude: windowPeriod, unit: 'ms'}],
-      },
-    },
-  ]
-}
-
-export const getWindowVars = (
-  query: string,
-  variables: VariableAssignment[]
-): VariableAssignment[] => {
-  if (!query.includes(WINDOW_PERIOD)) {
-    return []
-  }
-
-  const windowPeriod =
-    getWindowPeriod(query, variables) || FALLBACK_WINDOW_PERIOD
-
-  return [
-    {
-      type: 'VariableAssignment',
-      id: {
-        type: 'Identifier',
-        name: WINDOW_PERIOD,
-      },
-      init: {
-        type: 'DurationLiteral',
-        values: [{magnitude: windowPeriod, unit: 'ms'}],
-      },
-    },
-  ]
-}
-
-export const calcWindowPeriodForDuration = (queryDuration: number) =>
-  Math.round(queryDuration / DESIRED_POINTS_PER_GRAPH) // queryDuration in ms
-
-/*
-  Compute the duration (in milliseconds) to use for the `v.windowPeriod`
-  variable assignment for a query.
-*/
-export const getWindowPeriod = (
-  query: string,
-  variables: VariableAssignment[]
 ): number | null => {
   if (query.length === 0) {
     return null
   }
-  try {
-    const ast = parse(query)
-    const substitutedAST: Package = {
-      package: '',
-      type: 'Package',
-      files: [ast, buildVarsOption(variables)],
-    }
-
-    const queryDuration = getMinDurationFromAST(substitutedAST) // in ms
-
-    const foundDuration = SELECTABLE_TIME_RANGES.find(
-      tr => tr.seconds * 1000 === queryDuration
-    )
-
-    if (foundDuration) {
-      return foundDuration.windowPeriod
-    }
-
-    return calcWindowPeriodForDuration(queryDuration)
-  } catch (error) {
-    return null
-  }
-}
 
-export const getWindowPeriodFromVariables = (
-  query: string,
-  variables: Variable[]
-): number | null => {
-  if (query.length === 0) {
-    return null
-  }
   try {
-    const ast = parse(query)
-    const substitutedAST: Package = {
+    const varAssignments = variables
+      .map(v => asAssignmentNode(v))
+      .filter(v => !!v)
+
+    const outerScopeAst: Package = {
       package: '',
       type: 'Package',
-      files: [ast, buildUsedVarsOption(query, variables)],
-    }
-
-    const queryDuration = getMinDurationFromAST(substitutedAST) // in ms
-
-    const foundDuration = SELECTABLE_TIME_RANGES.find(
-      tr => tr.seconds * 1000 === queryDuration
-    )
-
-    if (foundDuration) {
-      return foundDuration.windowPeriod
+      files: [buildVarsOption(varAssignments)],
     }
 
-    return calcWindowPeriodForDuration(queryDuration)
+    const ast = parse(query)
+    return getDurationFromAST(ast, outerScopeAst)
   } catch (error) {
+    console.warn('error', error)
     return null
   }
 }
 
-export const getWindowPeriodVariable = (
+/*
+  Extract the UI abstraction Variable for `v.windowPeriod`
+*/
+export const getWindowPeriodVariableFromVariables = (
```

Comment:
* 2 main functions (in this file):
     * used for UI Variable
     * used for flux ast VariableAssignment.
* both now use the same core logic.